### PR TITLE
Supporting more BERT-like models

### PR DESCRIPTION
**Pull request by ashvardanian**
Hi HF team!

I am extending our UForm repository of multimodal models to support Swift and mobile deployments, and along that way I've noticed that several classes for a broad range of BERT-like models are not yet supported by `swift-transformers`. So I've added a `WordPieceDecoder` class and aliases for `BertPreTokenizer` and `BertProcessing`.

Moreover, are you are well aware `config.json` and `tokenizer.json` come in all shapes and sizes. So I've added fallback mechanisms to handle different tuple order in vocabulary listings.

The [current `main-dev` branch of UForm](https://github.com/unum-cloud/uform/commit/00bd84c59995c7b3daa0b4fa1597f77608806fdb) is already using this functionality from my fork. I am looking into integrating more `Hub` functionality next. Please let me know what you think about this PR 🤗

### CODE REVIEW

**Comment by ashvardanian** [comment]
Another open problem that I've recently discovered is the way strings are compared in Swift. By default, the language uses UTF8-aware normalization techniques when comparing strings. This is great for some applications, but horrible for tokenization, especially with multilingual models. I've solved that by introducing a `LiteralString` wrapper for `String`, that uses the literal comparators:

```swift
    struct LiteralString: Hashable {
        let value: String

        static func ==(lhs: LiteralString, rhs: LiteralString) -> Bool {
            return lhs.value.compare(rhs.value, options: .literal) == .orderedSame
        }

        func hash(into hasher: inout Hasher) {
            hasher.combine(value)
        }
    }
```

I believe it should be applicable in other places as well. Let me know what you think, @pcuenca 🤗

File: Sources/Tokenizers/PostProcessor.swift [hunk 1]
```diff
@@ -22,20 +22,30 @@ extension PostProcessor {
...
-    case TemplateProcessing
+    case Template
     case ByteLevel
-    case RobertaProcessing
+    case Bert
+    case Roberta
+
+    static let BertProcessing = "Bert"
+    static let RobertaProcessing = "Roberta"
+    static let TemplateProcessing = "Template"
```
Comment (pcuenca) [inline Sources/Tokenizers/PostProcessor.swift:32]: I'm not sure this is worth doing, I find the aliases and the suffix removal distracting for little benefit. My original approach was to simply use the same names that appear in the json, so someone reading both could easily match.

File: Sources/Tokenizers/UnigramTokenizer.swift [hunk 2]
```diff
@@ -50,11 +65,14 @@ class UnigramTokenizer: PreTrainedTokenizerModel {
         self.unknownTokenId = unknownTokenId
         self.unknownPiece = SentencePieceToken(token: vocab[unknownTokenId].token, score: minScore - 10)
         
-        tokensToIds = Dictionary(uniqueKeysWithValues: vocab.map { $0.token }.enumerated().map { ($1, $0) })
-        bosTokenId = tokensToIds[bosToken!]      // May be nil
+        // Using `Dictionary(uniqueKeysWithValues:)` is the default approach for constructing the mapping.
+        // It, however, will use the normal `compare` function of strings and will result in collisions for different 
+        // UTF8 strings. Instead, we should use the `a.compare(b, options: .literal) == .orderedSame`.
+        tokensToIds = Dictionary(uniqueKeysWithValues: vocab.map { $0.token }.enumerated().map { (LiteralString(value: $1), $0) })
+        bosTokenId = tokensToIds[LiteralString(value: bosToken!)]      // May be nil
```
Comment (pcuenca) [inline Sources/Tokenizers/UnigramTokenizer.swift:72]: maybe worth considering a `String` extension that returns a `LiteralString`, for readability? (Just an idea)

Comment (ashvardanian) [reply Sources/Tokenizers/UnigramTokenizer.swift:72]: Maybe 🤔

File: Sources/Hub/Hub.swift [hunk 3]
```diff
@@ -100,7 +100,19 @@ public struct Config {
...
+        }
+
+        if let first = array[0] as? String, let second = array[1] as? Int64 {
+            return (UInt(second), first)
+        } else if let first = array[0] as? Int64, let second = array[1] as? String {
+            return (UInt(first), second)
+        }
+
+        return nil
+    }
```
[the remark this answers is not in the document]
Comment (ashvardanian) [reply Sources/Hub/Hub.swift:115]: Sure, I was testing everything on our models, and we generally don't invent new config structures and reuse existing variants.

image

We are preparing a [release of the re-packaged encoders](https://huggingface.co/collections/unum-cloud/uform-3-encoders-662344ee44bc96a4aa5937cd). They work fine with Python and JavaScript Transformers libraries, but required all of these changes to work in Swift 🤗

File: Sources/Tokenizers/PostProcessor.swift (hunk 1 above)
Comment (ashvardanian) [reply Sources/Tokenizers/PostProcessor.swift:32]: It's very hard to use JSON as the ground truth, as they cone in all shapes and sizes. 

Shorter name was needed for our models to work, but I've added the static variables for backward-compatibility, to avoid breaking the library for other users.

File: Sources/Hub/Hub.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply Sources/Hub/Hub.swift:115]: Yes, I understand! It was for documentation purposes only (maybe as a comment if appropriate) :) But it's ok!

File: Sources/Tokenizers/PostProcessor.swift (hunk 1 above)
Comment (pcuenca) [reply Sources/Tokenizers/PostProcessor.swift:32]: I'd say the json names used in transformers should be quite stable right now. Why do your models require shorter names? (just curious)

Comment (ashvardanian) [reply Sources/Tokenizers/PostProcessor.swift:32]: I guess no model "requires" specific names, and JSONs can always be changed, but it generally results in a snowballing set of changes that have to be applied on every platform...

In UForm I have identical tests that run for the same models on the same data across all 3 languages across ONNX, PyTorch, and CoreML. You can [check them here](https://github.com/ashvardanian/uform/blob/main/CONTRIBUTING.md).

If a certain behavior is standard in the more popular ports of the library (Python and JS), I assume Hugging Face may want to provide the same behavior here to encourage adoption. A lot of people would probably appreciate the portability 🤗

**Comment by ConfuseIous** [comment]
Any chance of this being merged in soon? I'm trying to use a BERT model and this PR would be a huge help :)

**Comment by eemilk** [comment]
I was trying out these changes since have DistilBERT model I wanted to test, but getting thrown `unsupportedTokenizer("Distilbert")` on `try await AutoTokenizer.from(pretrained:..`

Is there a chance support for DistilBERT is also added?

**Comment by pcuenca** [comment]
Superseded by #137
